topostats/grainstats.py

```diff
@@ -0,0 +1,653 @@
+        ----------
+        point_2 : tuple
+            The point to find the squared distance to.
+        point_1 : tuple
+            Optional - defaults to the starting point defined in the graham_scan() function. The point to find the
+        squared distance from.
+
+        Returns
+        -------
+        distance_squared : float
+            The squared distance between the two points.
+        """
+        # Get the distance squared between two points. If the second point is not provided, use the starting point.
+        point_1 = self.start_point if point_1 == None else point_1
+        delta_x = point_2[0] - point_1[0]
+        delta_y = point_2[1] - point_1[1]
+        # Don't need the sqrt since just sorting for dist
+        return float(delta_x**2 + delta_y**2)
+
+    def sort_points(self, points: List) -> List:
+        #    def sort_points(self, points: np.array) -> List:
+        """Function to sort points in counter-clockwise order of angle made with the starting point.
+
+        Parameters
+        ----------
+        points: list
+            A python list of the coordinates to sort.
+
+        Returns
+        -------
+        sorted_points : list
+            A python list of sorted points.
+        """
+        # Return if the list is length 1 or 0 (i.e. a single point).
+        if len(points) <= 1:
+            return points
+        # Lists that allow sorting of points relative to a current comparision point
+        smaller, equal, larger = [], [], []
+        # Get a random point in the array to calculate the pivot angle from. This sorts the points relative to this point.
+        pivot_angle = self.get_angle(points[randint(0, len(points) - 1)], self.start_point)
+        for point in points:
+            point_angle = self.get_angle(point, self.start_point)
+            # If the
+            if point_angle < pivot_angle:
+                smaller.append(point)
+            elif point_angle == pivot_angle:
+                equal.append(point)
+            else:
+                larger.append(point)
+        # Lets take a different approach and use arrays, we have a start point lets work out the angle of each point
+        # relative to that and _then_ sort it.
+        # pivot_angles = self.get_angle(points, self.start_point)
+        # Recursively sort the arrays until each point is sorted
+        sorted_points = self.sort_points(smaller) + sorted(equal, key=self.get_displacement) + self.sort_points(larger)
+        # Return sorted array where equal angle points are sorted by distance
+        return sorted_points
+
+    def get_start_point(self, edges) -> int:
+        """Determine the index of the bottom most point of the hull when sorted by x-position.
+
+        Parameters
+        ----------
+        edges: np.array
+
+        Returns
+        -------
+        """
+        min_y_index = np.argmin(edges[:, 1])
+        self.start_point = edges[min_y_index]
+
+    def graham_scan(self, edges: list):
```

Comment:
Is this unit tested anywhere?

Reply:
For this and `convex_hull()` function I couldn't work out a sensible test and think both are candidates for refactoring into smaller methods (some of which I've done and in linting the code this morning pulled out another section of code). I was wary of going too far though as it was known there would be changes.

I'll create an Issue for outstanding code that needs unit-tests.